wdl-engine/src/value.rs

```diff
@@ -833,14 +1036,62 @@ impl PrimitiveValue {
 
         Display(self)
     }
+
+    /// Visits each file or directory path contained in the value.
+    fn visit_paths(&self, cb: &mut impl FnMut(&str)) {
+        match self {
+            Self::File(path) | Self::Directory(path) => cb(path.as_str()),
+            _ => {}
+        }
+    }
+
+    /// Replaces any inner path values by joining the specified path with the
+    /// path value.
+    ///
+    /// If `check_existence` is `true` and a required path does not exist, an
+    /// error is returned. An optional path that does not exist is replaced with
+    /// `None`.
+    fn join_paths(&mut self, path: &Path, check_existence: bool, optional: bool) -> Result<bool> {
+        match self {
+            PrimitiveValue::File(p) => {
+                if let Ok(joined) = path.join(p.as_str()).into_os_string().into_string() {
+                    *Arc::make_mut(p) = joined;
+                }
+
+                if check_existence && !Path::new(p.as_str()).is_file() {
+                    if optional {
+                        return Ok(false);
+                    } else {
+                        bail!("file `{p}` does not exist");
+                    }
+                }
+            }
+            PrimitiveValue::Directory(p) => {
+                if let Ok(joined) = path.join(p.as_str()).into_os_string().into_string() {
+                    *Arc::make_mut(p) = joined;
+                }
+
+                if check_existence && !Path::new(p.as_str()).is_dir() {
+                    if optional {
+                        return Ok(false);
+                    } else {
+                        bail!("directory `{p}` does not exist");
+                    }
+                }
+            }
+            _ => {}
+        }
+
+        Ok(true)
+    }
 }
 
 impl fmt::Display for PrimitiveValue {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
             Self::Boolean(v) => write!(f, "{v}"),
             Self::Integer(v) => write!(f, "{v}"),
-            Self::Float(v) => write!(f, "{v:?}"),
+            Self::Float(v) => write!(f, "{v:.6?}"),
```

Comment:
Ah, I see it's intentional here. Could you share the thought process behind being explicit here?